glean-core/src/event_database/mod.rs

```diff
@@ -0,0 +1,375 @@
+        RecordedEventData::serialize_parts(
+            self.timestamp - timestamp_offset,
+            &self.category,
+            &self.name,
+            &self.extra,
+        )
+    }
+
+    /// Internal function to perform the serialization to JSON
+    fn serialize_parts(
+        timestamp: u64,
+        category: &str,
+        name: &str,
+        extra: &Option<HashMap<String, String>>,
+    ) -> JsonValue {
+        let mut result = JsonMap::new();
+        result.insert(TIMESTAMP_FIELD.to_string(), json!(timestamp));
+        result.insert(CATEGORY_FIELD.to_string(), json!(category));
+        result.insert(NAME_FIELD.to_string(), json!(name));
+        if let Some(extra) = extra {
+            result.insert(
+                EXTRA_FIELD.to_string(),
+                JsonValue::Object(
+                    extra
+                        .iter()
+                        .map(|(k, v)| (k.clone(), json!(v)))
+                        .collect::<JsonMap<String, JsonValue>>(),
+                ),
+            );
+        }
+        JsonValue::Object(result)
+    }
+
+    /// Deserialize an event from JSON
+    pub fn deserialize(value: &JsonValue) -> Option<RecordedEventData> {
+        let extra = value
+            .get(EXTRA_FIELD.to_string())
+            .and_then(|extra| extra.as_object())
+            .and_then(|extra| {
+                Some(
+                    extra
+                        .iter()
+                        .filter(|(_, v)| v.as_str().is_some())
+                        .map(|(k, v)| (k.to_string(), v.as_str().unwrap().to_string()))
+                        .collect(),
+                )
+            });
+        Some(RecordedEventData {
+            timestamp: value.get(TIMESTAMP_FIELD.to_string())?.as_u64()?,
+            category: value.get(CATEGORY_FIELD.to_string())?.as_str()?.to_string(),
+            name: value.get(NAME_FIELD.to_string())?.as_str()?.to_string(),
+            extra,
+        })
+    }
+}
+
+/// This struct handles the in-memory and on-disk storage logic for events.
+///
+/// So that the data survives shutting down of the application, events are stored
+/// in an append-only file on disk, in addition to the store in memory. Each line
+/// of this file records a single event in JSON, exactly as it will be sent in the
+/// ping.  There is one file per store.
+///
+/// When restarting the application, these on-disk files are checked, and if any are
+/// found, they are loaded, queued for sending and flushed immediately before any
+/// further events are collected.  This is because the timestamps for these events
+/// may have come from a previous boot of the device, and therefore will not be
+/// compatible with any newly-collected events.
+#[derive(Debug)]
+pub struct EventDatabase {
+    pub path: PathBuf,
+    // The in-memory list of events
+    event_stores: RwLock<HashMap<String, Vec<RecordedEventData>>>,
+}
+
+impl EventDatabase {
```

Comment:
Do we really need this, here? Maybe we could leverage the database we already have for this? We used this approach in glean-ac mostly to cope with crashes and to make sure we were not loosing events. But do we have the same risk here, given the storage we're using now, which should be a bit more resilient?

@badboy thoughts?

Reply:
(Some of this was discussed in the meeting, but wanted to summarize here for the record)...

rkv doesn't seem to have a good way to handle data that is an appended list of entries, rather than key/values.  We *could* put the index in the key if we wanted to, but just figuring out what the next index should be when adding an event would be a non-constant operation.

Storing events on disk is required not just for crashes -- the OS routinely shuts down the app without warning even if it is behaving correctly, so we need to persist (through some mechanism) regardless.

A possible simplification (and the approach taken in `service-telemetry`) is to *only* store events on disk.  That was [reviewed out](https://github.com/mozilla-mobile/android-components/pull/2328#pullrequestreview-213441306) in the glean-ac implementation since it would require I/O on the same thread as other Glean operations (but never on the main thread).  In glean-core, we don't fire off I/O threads anyway -- maybe we should, but if not, that simplification might make a lot of sense here and significantly reduce the complexity of the implementation.

Reply:
It occurs to me that if we wanted to use a robust data store that meets the requirements, we could use `sqlite`, but that would be an extra dependency.  It ships on all Android devices, so not *really* a new dependency, but the Rust wrapper would need to be configured to use it, etc., which could be thorny.

Reply:
Yeah, plus we don't know how the story looks on iOS yet. It's probably not worth the effort :)

Reply:
FxA has wrappers around sqlite, but it's a bit more complicated and I don't see the need in bringing in another database when the filesystem ought to be good enough for now.